### PR TITLE
fix(oscontext): Improvements to Unity OS context parsing [INGEST-385]

### LINKED ISSUE #1088: OS parsing issues for iOS, Android, Linux
After iOS 15 update the raw_description sent is:

```
    "os": {
      "raw_description": "iOS 15.0",
      "type": "os"
    },
```

image

And Android: ([issue](https://sentry.io/organizations/sentry-sdks/issues/2671354296/?project=5439417&query=is%3Aunresolved))

`Android OS 4.4.2 / API-19 (KOT49H/A536_S186_150813_ROW)`

Doesn't seem to get parsed by relay

Ubuntu example: `Linux 5.11 Ubuntu 20.04 64bit`

### PR DESCRIPTION
**Pull request by untitaker**
Fix #1088

### CODE REVIEW

**Comment by marandaneto** [comment]
for Android its very inconsistent depending on the dist.

Screenshot 2021-12-09 at 14 51 50

`os_linux` is added by sentry-native since it's an NDK issue.

File: relay-general/src/store/normalize/contexts.rs [hunk 1]
```diff
@@ -8,14 +8,17 @@ lazy_static::lazy_static! {
     static ref OS_WINDOWS_REGEX1: Regex = Regex::new(r#"^(Microsoft )?Windows (NT )?(?P<version>\d+\.\d+\.\d+).*$"#).unwrap();
     static ref OS_WINDOWS_REGEX2: Regex = Regex::new(r#"^Windows\s+\d+\s+\((?P<version>\d+\.\d+\.\d+).*$"#).unwrap();
 
-    static ref OS_ANDROID_REGEX: Regex = Regex::new(r#"^Android OS (?P<version>(\d+)) / API-(?P<api>(\d+))"#).unwrap();
+    static ref OS_ANDROID_REGEX: Regex = Regex::new(r#"^Android (OS )?(?P<version>\d+(\.\d+){0,2}) / API-(?P<api>(\d+))"#).unwrap();
```
Comment (marandaneto) [inline relay-general/src/store/normalize/contexts.rs:11]: isn't a breaking change? I believe we can see both use cases here, `Android OS...` and `Android (OS ...)` with or without `()`

Comment (untitaker) [reply relay-general/src/store/normalize/contexts.rs:11]: This regex accepts strictly more values than before is all I can say. What is a breaking change in this context?

**Comment by untitaker** [comment]
@marandaneto 

>for Android its very inconsistent depending on the dist.

Where is this screenshot from? Is this tested with this Relay? What is the raw_description input here?

**Comment by marandaneto** [comment]
> @marandaneto
> 
> > for Android its very inconsistent depending on the dist.
> 
> Where is this screenshot from? Is this tested with this Relay? What is the raw_description input here?

no, current relay, it was just to show that on Android and its fragmentation, the regex will likely not work, its not to be a blocker to this PR, just sharing the info.
https://sentry.io/organizations/sentry-sdks/issues/2736937471/?project=5428559&query=is%3Aunresolved

File: relay-general/src/store/normalize/contexts.rs (hunk 1 above)
Comment (marandaneto) [reply relay-general/src/store/normalize/contexts.rs:11]: if its more inclusive, all good.
Regex isn't my best skill.

Comment (untitaker) [reply relay-general/src/store/normalize/contexts.rs:11]: It now accepts both `Android ` and `Android OS `, then also some extra segments in the version number

**Review by bruno-garcia — APPROVED** [review]

File: relay-general/src/store/normalize/contexts.rs [hunk 2]
```diff
@@ -112,10 +115,20 @@ fn normalize_os_context(os: &mut OsContext) {
                 .name("build")
                 .map(|m| m.as_str().to_string().into())
                 .into();
+        } else if let Some(captures) = OS_LINUX_DISTRO_UNAME_REGEX.captures(raw_description) {
```
Comment (jjbayer) [inline relay-general/src/store/normalize/contexts.rs:118]: nit: Do we have any stats on what OS is the most frequent? If Linux is much more frequent than MacOS, should we attempt to parse it first?

**Review by jjbayer — APPROVED** [review]

File: relay-general/src/store/normalize/contexts.rs (hunk 2 above)
Comment (untitaker) [reply relay-general/src/store/normalize/contexts.rs:118]: yeah no idea, i think it won't matter. We should find a way to merge all regexes into one, ideally.

**Pull request merged**